File: http/http-client-annotation-processor/src/main/java/ru/tinkoff/kora/http/client/annotation/processor/extension/HttpClientKoraExtension.java
package ru.tinkoff.kora.http.client.annotation.processor.extension;

import ru.tinkoff.kora.annotation.processor.common.CommonUtils;
import ru.tinkoff.kora.http.client.annotation.processor.HttpClientAnnotationProcessor;
import ru.tinkoff.kora.http.client.annotation.processor.HttpClientUtils;
import ru.tinkoff.kora.http.client.common.annotation.HttpClient;
import ru.tinkoff.kora.kora.app.annotation.processor.extension.ExtensionResult;
import ru.tinkoff.kora.kora.app.annotation.processor.extension.KoraExtension;

import javax.annotation.Nullable;
import javax.annotation.processing.ProcessingEnvironment;
import javax.annotation.processing.RoundEnvironment;
import javax.lang.model.element.ElementKind;
import javax.lang.model.element.ExecutableElement;
import javax.lang.model.element.TypeElement;
import javax.lang.model.type.TypeMirror;
import javax.lang.model.util.Elements;
import javax.lang.model.util.Types;

public class HttpClientKoraExtension implements KoraExtension {
    private final Elements elements;
    private final Types types;
    private final ProcessingEnvironment processingEnv;
    private final HttpClientAnnotationProcessor processor;

    public HttpClientKoraExtension(ProcessingEnvironment processingEnvironment) {
        this.processingEnv = processingEnvironment;
        this.elements = processingEnvironment.getElementUtils();
        this.types = processingEnvironment.getTypeUtils();
        this.processor = new HttpClientAnnotationProcessor();
        this.processor.init(processingEnvironment);
    }

    @Nullable
    @Override
    public KoraExtensionDependencyGenerator getDependencyGenerator(RoundEnvironment roundEnvironment, TypeMirror typeMirror) {
        var element = this.types.asElement(typeMirror);
        if (element == null || element.getKind() != ElementKind.INTERFACE) {
            return null;
        }
        var annotation = element.getAnnotation(HttpClient.class);
        if (annotation == null) {
            return null;
        }
        return () -> {
            var typeElement = (TypeElement) element;
            var implName = HttpClientUtils.clientName(typeElement);
            var packageName = this.processingEnv.getElementUtils().getPackageOf(typeElement).getQualifiedName().toString();
            var maybeGenerated = this.elements.getTypeElement(packageName + "." + implName);
            if (maybeGenerated == null) {
                // annotation processor will handle it
                return ExtensionResult.nextRound();
            }
            var aopProxy = CommonUtils.getOuterClassesAsPrefix(maybeGenerated) + maybeGenerated.getSimpleName() + "__AopProxy";
            var aopProxyElement = this.elements.getTypeElement(packageName + "." + aopProxy);
            if (aopProxyElement == null) {
                // aop annotation processor will handle it
                return ExtensionResult.nextRound();
            }
            return maybeGenerated.getEnclosedElements().stream()
                .filter(e -> e.getKind() == ElementKind.CONSTRUCTOR)
                .map(ExecutableElement.class::cast)
                .map(ExtensionResult::fromExecutable)
                .findFirst()
                .orElseThrow();

        };
    }
}

File: http/http-client-symbol-processor/src/main/kotlin/ru/tinkoff/kora/http/client/symbol/processor/extension/HttpClientKoraExtension.kt
package ru.tinkoff.kora.http.client.symbol.processor.extension

import com.google.devtools.ksp.KspExperimental
import com.google.devtools.ksp.getAnnotationsByType
import com.google.devtools.ksp.getClassDeclarationByName
import com.google.devtools.ksp.getConstructors
import com.google.devtools.ksp.processing.Resolver
import com.google.devtools.ksp.symbol.ClassKind
import com.google.devtools.ksp.symbol.KSClassDeclaration
import com.google.devtools.ksp.symbol.KSType
import ru.tinkoff.kora.http.client.common.annotation.HttpClient
import ru.tinkoff.kora.http.client.symbol.processor.clientName
import ru.tinkoff.kora.kora.app.ksp.extension.ExtensionResult
import ru.tinkoff.kora.kora.app.ksp.extension.KoraExtension
import ru.tinkoff.kora.ksp.common.CommonAopUtils.hasAopAnnotations
import ru.tinkoff.kora.ksp.common.getOuterClassesAsPrefix

@KspExperimental
class HttpClientKoraExtension : KoraExtension {
    override fun getDependencyGenerator(resolver: Resolver, type: KSType): (() -> ExtensionResult)? {
        val declaration = type.declaration
        if (declaration !is KSClassDeclaration || declaration.classKind != ClassKind.INTERFACE) {
            return null
        }
        if (declaration.getAnnotationsByType(HttpClient::class).firstOrNull() == null) {
            return null
        }
        return lambda@{
            val implName = declaration.clientName()
            val packageName: String = declaration.packageName.asString()
            val maybeGenerated = resolver.getClassDeclarationByName("$packageName.$implName")
            if (maybeGenerated == null) {
                return@lambda ExtensionResult.RequiresCompilingResult
            }
            if (!hasAopAnnotations(resolver, maybeGenerated)) {
                return@lambda maybeGenerated.getConstructors().map { ExtensionResult.fromConstructor(it, maybeGenerated) }.first()
            }
            val aopProxy = maybeGenerated.getOuterClassesAsPrefix() + maybeGenerated.simpleName.getShortName() + "__AopProxy"
            val aopProxyElement = resolver.getClassDeclarationByName("$packageName.$aopProxy")
            if (aopProxyElement == null) {
                return@lambda ExtensionResult.RequiresCompilingResult
            }
            return@lambda aopProxyElement.getConstructors().map { ExtensionResult.fromConstructor(it, aopProxyElement) }.first()
        }
    }
}

